File: app/src/main/java/com/voltpay/app/data/db/ContactDao.java
package com.voltpay.app.data.db;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import net.zetetic.database.sqlcipher.SQLiteDatabase;
import com.voltpay.app.data.model.Contact;

import java.util.ArrayList;
import java.util.List;

public class ContactDao {
    private final DatabaseHelper dbHelper;

    public ContactDao(Context context) {
        this.dbHelper = DatabaseHelper.getInstance(context);
    }

    public long insertContact(Contact contact) {
        SQLiteDatabase db = dbHelper.getSecureWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(DatabaseHelper.COL_CONTACT_NAME, contact.getName());
        values.put(DatabaseHelper.COL_CONTACT_UPI_ID, contact.getUpiId());
        values.put(DatabaseHelper.COL_CONTACT_PHONE, contact.getPhoneNumber());
        values.put(DatabaseHelper.COL_CONTACT_CREATED_AT, contact.getCreatedAt());
        values.put(DatabaseHelper.COL_CONTACT_LAST_PAID_AT, contact.getLastPaidAt());
        
        long id = db.insertWithOnConflict(DatabaseHelper.TABLE_CONTACTS, null, values, SQLiteDatabase.CONFLICT_REPLACE);
        db.close();
        return id;
    }

    public List<Contact> getAllContacts() {
        List<Contact> contacts = new ArrayList<>();
        SQLiteDatabase db = dbHelper.getSecureReadableDatabase();
        Cursor cursor = db.query(DatabaseHelper.TABLE_CONTACTS, null, null, null, null, null, DatabaseHelper.COL_CONTACT_NAME + " ASC");
        
        if (cursor.moveToFirst()) {
            do {
                Contact contact = new Contact();
                contact.setId(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_ID)));
                contact.setName(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_NAME)));
                contact.setUpiId(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_UPI_ID)));
                contact.setPhoneNumber(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_PHONE)));
                contact.setCreatedAt(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_CREATED_AT)));
                contact.setLastPaidAt(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_LAST_PAID_AT)));
                contacts.add(contact);
            } while (cursor.moveToNext());
        }
        cursor.close();
        db.close();
        return contacts;
    }

    public int updateContact(Contact contact) {
        SQLiteDatabase db = dbHelper.getSecureWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(DatabaseHelper.COL_CONTACT_NAME, contact.getName());
        values.put(DatabaseHelper.COL_CONTACT_UPI_ID, contact.getUpiId());
        values.put(DatabaseHelper.COL_CONTACT_PHONE, contact.getPhoneNumber());
        values.put(DatabaseHelper.COL_CONTACT_LAST_PAID_AT, contact.getLastPaidAt());
        
        int rows = db.update(DatabaseHelper.TABLE_CONTACTS, values, DatabaseHelper.COL_CONTACT_ID + " = ?", new String[]{String.valueOf(contact.getId())});
        db.close();
        return rows;
    }

    public int deleteContact(long id) {
        SQLiteDatabase db = dbHelper.getSecureWritableDatabase();
        int rows = db.delete(DatabaseHelper.TABLE_CONTACTS, DatabaseHelper.COL_CONTACT_ID + " = ?", new String[]{String.valueOf(id)});
        db.close();
        return rows;
    }

    public Contact getById(long id) {
        SQLiteDatabase db = dbHelper.getSecureReadableDatabase();
        Cursor cursor = db.query(DatabaseHelper.TABLE_CONTACTS, null, DatabaseHelper.COL_CONTACT_ID + " = ?", new String[]{String.valueOf(id)}, null, null, null);
        Contact contact = null;
        if (cursor.moveToFirst()) {
            contact = new Contact();
            contact.setId(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_ID)));
            contact.setName(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_NAME)));
            contact.setUpiId(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_UPI_ID)));
            contact.setPhoneNumber(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_PHONE)));
            contact.setCreatedAt(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_CREATED_AT)));
            contact.setLastPaidAt(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_LAST_PAID_AT)));
        }
        cursor.close();
        db.close();
        return contact;
    }

    public Contact getContactByUpiId(String upiId) {
        SQLiteDatabase db = dbHelper.getSecureReadableDatabase();
        Cursor cursor = db.query(DatabaseHelper.TABLE_CONTACTS, null, DatabaseHelper.COL_CONTACT_UPI_ID + " = ?", new String[]{upiId}, null, null, null);
        Contact contact = null;
        if (cursor.moveToFirst()) {
            contact = new Contact();
            contact.setId(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_ID)));
            contact.setName(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_NAME)));
            contact.setUpiId(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_UPI_ID)));
            contact.setPhoneNumber(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_PHONE)));
            contact.setCreatedAt(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_CREATED_AT)));
            contact.setLastPaidAt(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_LAST_PAID_AT)));
        }
        cursor.close();
        db.close();
        return contact;
    }

    public void updateLastPaidAt(String upiId, long timestamp) {
        SQLiteDatabase db = dbHelper.getSecureWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(DatabaseHelper.COL_CONTACT_LAST_PAID_AT, timestamp);
        db.update(DatabaseHelper.TABLE_CONTACTS, values, DatabaseHelper.COL_CONTACT_UPI_ID + " = ?", new String[]{upiId});
        db.close();
    }

    public List<Contact> searchByName(String query) {
        List<Contact> contacts = new ArrayList<>();
        SQLiteDatabase db = dbHelper.getSecureReadableDatabase();
        Cursor cursor = db.query(DatabaseHelper.TABLE_CONTACTS, null, DatabaseHelper.COL_CONTACT_NAME + " LIKE ?", new String[]{"%" + query + "%"}, null, null, DatabaseHelper.COL_CONTACT_NAME + " ASC");
        if (cursor.moveToFirst()) {
            do {
                Contact contact = new Contact();
                contact.setId(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_ID)));
                contact.setName(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_NAME)));
                contact.setUpiId(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_UPI_ID)));
                contact.setPhoneNumber(cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_PHONE)));
                contact.setCreatedAt(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_CREATED_AT)));
                contact.setLastPaidAt(cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_LAST_PAID_AT)));
                contacts.add(contact);
            } while (cursor.moveToNext());
        }
        cursor.close();
        db.close();
        return contacts;
    }
}

File: app/src/main/java/com/voltpay/app/data/db/ContactDao.kt
package com.voltpay.app.data.db

import android.content.ContentValues
import android.content.Context
import com.voltpay.app.data.model.Contact
import net.zetetic.database.sqlcipher.SQLiteDatabase
import java.util.ArrayList

class ContactDao(context: Context) {
    private val dbHelper: DatabaseHelper = DatabaseHelper.getInstance(context)

    fun insertContact(contact: Contact): Long {
        val db = dbHelper.secureWritableDatabase
        val values = ContentValues().apply {
            put(DatabaseHelper.COL_CONTACT_NAME, contact.name)
            put(DatabaseHelper.COL_CONTACT_UPI_ID, contact.upiId)
            put(DatabaseHelper.COL_CONTACT_PHONE, contact.phoneNumber)
            put(DatabaseHelper.COL_CONTACT_CREATED_AT, contact.createdAt)
            put(DatabaseHelper.COL_CONTACT_LAST_PAID_AT, contact.lastPaidAt)
        }
        val id = db.insertWithOnConflict(DatabaseHelper.TABLE_CONTACTS, null, values, SQLiteDatabase.CONFLICT_REPLACE)
        db.close()
        return id
    }

    fun getAllContacts(): List<Contact> {
        val contacts = ArrayList<Contact>()
        val db = dbHelper.secureReadableDatabase
        val cursor = db.query(DatabaseHelper.TABLE_CONTACTS, null, null, null, null, null, DatabaseHelper.COL_CONTACT_NAME + " ASC")
        
        if (cursor.moveToFirst()) {
            do {
                val contact = Contact()
                contact.id = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_ID))
                contact.name = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_NAME))
                contact.upiId = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_UPI_ID))
                contact.phoneNumber = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_PHONE))
                contact.createdAt = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_CREATED_AT))
                contact.lastPaidAt = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_LAST_PAID_AT))
                contacts.add(contact)
            } while (cursor.moveToNext())
        }
        cursor.close()
        db.close()
        return contacts
    }

    fun updateContact(contact: Contact): Int {
        val db = dbHelper.secureWritableDatabase
        val values = ContentValues().apply {
            put(DatabaseHelper.COL_CONTACT_NAME, contact.name)
            put(DatabaseHelper.COL_CONTACT_UPI_ID, contact.upiId)
            put(DatabaseHelper.COL_CONTACT_PHONE, contact.phoneNumber)
            put(DatabaseHelper.COL_CONTACT_LAST_PAID_AT, contact.lastPaidAt)
        }
        val rows = db.update(DatabaseHelper.TABLE_CONTACTS, values, DatabaseHelper.COL_CONTACT_ID + " = ?", arrayOf(contact.id.toString()))
        db.close()
        return rows
    }

    fun deleteContact(id: Long): Int {
        val db = dbHelper.secureWritableDatabase
        val rows = db.delete(DatabaseHelper.TABLE_CONTACTS, DatabaseHelper.COL_CONTACT_ID + " = ?", arrayOf(id.toString()))
        db.close()
        return rows
    }

    fun getById(id: Long): Contact? {
        val db = dbHelper.secureReadableDatabase
        val cursor = db.query(DatabaseHelper.TABLE_CONTACTS, null, DatabaseHelper.COL_CONTACT_ID + " = ?", arrayOf(id.toString()), null, null, null)
        var contact: Contact? = null
        if (cursor.moveToFirst()) {
            contact = Contact()
            contact.id = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_ID))
            contact.name = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_NAME))
            contact.upiId = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_UPI_ID))
            contact.phoneNumber = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_PHONE))
            contact.createdAt = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_CREATED_AT))
            contact.lastPaidAt = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_LAST_PAID_AT))
        }
        cursor.close()
        db.close()
        return contact
    }

    fun getContactByUpiId(upiId: String): Contact? {
        val db = dbHelper.secureReadableDatabase
        val cursor = db.query(DatabaseHelper.TABLE_CONTACTS, null, DatabaseHelper.COL_CONTACT_UPI_ID + " = ?", arrayOf(upiId), null, null, null)
        var contact: Contact? = null
        if (cursor.moveToFirst()) {
            contact = Contact()
            contact.id = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_ID))
            contact.name = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_NAME))
            contact.upiId = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_UPI_ID))
            contact.phoneNumber = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_PHONE))
            contact.createdAt = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_CREATED_AT))
            contact.lastPaidAt = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_LAST_PAID_AT))
        }
        cursor.close()
        db.close()
        return contact
    }

    fun updateLastPaidAt(upiId: String, timestamp: Long) {
        val db = dbHelper.secureWritableDatabase
        val values = ContentValues().apply {
            put(DatabaseHelper.COL_CONTACT_LAST_PAID_AT, timestamp)
        }
        db.update(DatabaseHelper.TABLE_CONTACTS, values, DatabaseHelper.COL_CONTACT_UPI_ID + " = ?", arrayOf(upiId))
        db.close()
    }

    fun searchByName(query: String): List<Contact> {
        val contacts = ArrayList<Contact>()
        val db = dbHelper.secureReadableDatabase
        val cursor = db.query(DatabaseHelper.TABLE_CONTACTS, null, DatabaseHelper.COL_CONTACT_NAME + " LIKE ?", arrayOf("%$query%"), null, null, DatabaseHelper.COL_CONTACT_NAME + " ASC")
        if (cursor.moveToFirst()) {
            do {
                val contact = Contact()
                contact.id = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_ID))
                contact.name = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_NAME))
                contact.upiId = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_UPI_ID))
                contact.phoneNumber = cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_PHONE))
                contact.createdAt = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_CREATED_AT))
                contact.lastPaidAt = cursor.getLong(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_CONTACT_LAST_PAID_AT))
                contacts.add(contact)
            } while (cursor.moveToNext())
        }
        cursor.close()
        db.close()
        return contacts
    }
}
